# Show the dataframe shape.

In [5]:
import pandas as pd
import numpy as np
marketing_data = pd.read_csv('/Users/joluhergo/Desktop/Ironhack/labs/lab-customer-analysis-r2/marketing_customer_analysis.csv')

In [6]:
marketing_data.shape

(10910, 26)

# Standardize header names.

In [7]:
marketing_data.columns

Index(['Unnamed: 0', 'Customer', 'State', 'Customer Lifetime Value',
       'Response', 'Coverage', 'Education', 'Effective To Date',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy Type', 'Policy', 'Renew Offer Type',
       'Sales Channel', 'Total Claim Amount', 'Vehicle Class', 'Vehicle Size',
       'Vehicle Type'],
      dtype='object')

In [9]:
marketing_data.columns = list(map(lambda x: x.upper(), marketing_data.columns))

In [12]:
marketing_data.columns

Index(['UNNAMED: 0', 'CUSTOMER', 'STATE', 'CUSTOMER LIFETIME VALUE',
       'RESPONSE', 'COVERAGE', 'EDUCATION', 'EFFECTIVE TO DATE',
       'EMPLOYMENTSTATUS', 'GENDER', 'INCOME', 'LOCATION CODE',
       'MARITAL STATUS', 'MONTHLY PREMIUM AUTO', 'MONTHS SINCE LAST CLAIM',
       'MONTHS SINCE POLICY INCEPTION', 'NUMBER OF OPEN COMPLAINTS',
       'NUMBER OF POLICIES', 'POLICY TYPE', 'POLICY', 'RENEW OFFER TYPE',
       'SALES CHANNEL', 'TOTAL CLAIM AMOUNT', 'VEHICLE CLASS', 'VEHICLE SIZE',
       'VEHICLE TYPE'],
      dtype='object')

# Which columns are numerical?

In [14]:
marketing_data.head(2)

UNNAMED: 0 CUSTOMER       STATE  CUSTOMER LIFETIME VALUE RESPONSE COVERAGE  \
0           0  DK49336     Arizona              4809.216960       No    Basic   
1           1  KX64629  California              2228.525238       No    Basic   

  EDUCATION EFFECTIVE TO DATE EMPLOYMENTSTATUS GENDER  ...  \
0   College           2/18/11         Employed      M  ...   
1   College           1/18/11       Unemployed      F  ...   

   NUMBER OF OPEN COMPLAINTS NUMBER OF POLICIES     POLICY TYPE        POLICY  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   

   RENEW OFFER TYPE  SALES CHANNEL  TOTAL CLAIM AMOUNT  VEHICLE CLASS  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   

  VEHICLE SIZE VEHICLE TYPE  
0      Medsize          NaN  
1      Medsize          NaN  

[2 rows x 26 columns]

In [15]:
marketing_data.dtypes

UNNAMED: 0                         int64
CUSTOMER                          object
STATE                             object
CUSTOMER LIFETIME VALUE          float64
RESPONSE                          object
COVERAGE                          object
EDUCATION                         object
EFFECTIVE TO DATE                 object
EMPLOYMENTSTATUS                  object
GENDER                            object
INCOME                             int64
LOCATION CODE                     object
MARITAL STATUS                    object
MONTHLY PREMIUM AUTO               int64
MONTHS SINCE LAST CLAIM          float64
MONTHS SINCE POLICY INCEPTION      int64
NUMBER OF OPEN COMPLAINTS        float64
NUMBER OF POLICIES                 int64
POLICY TYPE                       object
POLICY                            object
RENEW OFFER TYPE                  object
SALES CHANNEL                     object
TOTAL CLAIM AMOUNT               float64
VEHICLE CLASS                     object
VEHICLE SIZE    

In [17]:
n = ['int64','float64']
marketing_data.select_dtypes(include=n)

UNNAMED: 0  CUSTOMER LIFETIME VALUE  INCOME  MONTHLY PREMIUM AUTO  \
0               0              4809.216960   48029                    61   
1               1              2228.525238       0                    64   
2               2             14947.917300   22139                   100   
3               3             22332.439460   49078                    97   
4               4              9025.067525   23675                   117   
...           ...                      ...     ...                   ...   
10905       10905             15563.369440       0                   253   
10906       10906              5259.444853   61146                    65   
10907       10907             23893.304100   39837                   201   
10908       10908             11971.977650   64195                   158   
10909       10909              6857.519928       0                   101   

       MONTHS SINCE LAST CLAIM  MONTHS SINCE POLICY INCEPTION  \
0                          7.0                             52   
1                          3.0                             26   
2                         34.0                             31   
3                         10.0                              3   
4                          NaN                             31   
...                        ...                            ...   
10905                      NaN                             40   
10906                      7.0                             68   
10907                     11.0                             63   
10908                      0.0                             27   
10909                     31.0                              1   

       NUMBER OF OPEN COMPLAINTS  NUMBER OF POLICIES  TOTAL CLAIM AMOUNT  
0                            0.0                   9          292.800000  
1                            0.0                   1          744.924331  
2                            0.0                   2          480.000000  
3                            0.0                   2          484.013411  
4                            NaN                   7          707.925645  
...                          ...                 ...                 ...  
10905                        NaN                   7         1214.400000  
10906                        0.0                   6          273.018929  
10907                        0.0                   2          381.306996  
10908                        4.0                   6          618.288849  
10909                        0.0                   3         1021.719397  

[10910 rows x 9 columns]

# Which columns are categorical?

In [18]:
marketing_data.select_dtypes(include='object')

CUSTOMER       STATE RESPONSE  COVERAGE EDUCATION EFFECTIVE TO DATE  \
0      DK49336     Arizona       No     Basic   College           2/18/11   
1      KX64629  California       No     Basic   College           1/18/11   
2      LZ68649  Washington       No     Basic  Bachelor           2/10/11   
3      XL78013      Oregon      Yes  Extended   College           1/11/11   
4      QA50777      Oregon       No   Premium  Bachelor           1/17/11   
...        ...         ...      ...       ...       ...               ...   
10905  FE99816      Nevada       No   Premium  Bachelor           1/19/11   
10906  KX53892      Oregon       No     Basic   College            1/6/11   
10907  TL39050     Arizona       No  Extended  Bachelor            2/6/11   
10908  WA60547  California       No   Premium   College           2/13/11   
10909  IV32877         NaN      NaN     Basic  Bachelor            1/8/11   

      EMPLOYMENTSTATUS GENDER LOCATION CODE MARITAL STATUS     POLICY TYPE  \
0             Employed      M      Suburban        Married  Corporate Auto   
1           Unemployed      F      Suburban         Single   Personal Auto   
2             Employed      M      Suburban         Single   Personal Auto   
3             Employed      M      Suburban         Single  Corporate Auto   
4        Medical Leave      F      Suburban        Married   Personal Auto   
...                ...    ...           ...            ...             ...   
10905       Unemployed      F      Suburban        Married   Personal Auto   
10906         Employed      F         Urban        Married   Personal Auto   
10907         Employed      F         Rural        Married  Corporate Auto   
10908         Employed      F         Urban       Divorced   Personal Auto   
10909       Unemployed      M      Suburban         Single   Personal Auto   

             POLICY RENEW OFFER TYPE SALES CHANNEL  VEHICLE CLASS  \
0      Corporate L3           Offer3         Agent  Four-Door Car   
1       Personal L3           Offer4   Call Center  Four-Door Car   
2       Personal L3           Offer3   Call Center            SUV   
3      Corporate L3           Offer2        Branch  Four-Door Car   
4       Personal L2           Offer1        Branch  Four-Door Car   
...             ...              ...           ...            ...   
10905   Personal L1           Offer3           Web     Luxury Car   
10906   Personal L3           Offer2        Branch  Four-Door Car   
10907  Corporate L3           Offer1           Web     Luxury SUV   
10908   Personal L1           Offer1        Branch            SUV   
10909   Personal L1           Offer4           Web            SUV   

      VEHICLE SIZE VEHICLE TYPE  
0          Medsize          NaN  
1          Medsize          NaN  
2          Medsize            A  
3          Medsize            A  
4          Medsize          NaN  
...            ...          ...  
10905      Medsize            A  
10906      Medsize            A  
10907      Medsize          NaN  
10908      Medsize            A  
10909      Medsize          NaN  

[10910 rows x 17 columns]

# Check and deal with NaN values.

marketing_data.isna()

In [23]:
marketing_data.isna().sum()

UNNAMED: 0                          0
CUSTOMER                            0
STATE                             631
CUSTOMER LIFETIME VALUE             0
RESPONSE                          631
COVERAGE                            0
EDUCATION                           0
EFFECTIVE TO DATE                   0
EMPLOYMENTSTATUS                    0
GENDER                              0
INCOME                              0
LOCATION CODE                       0
MARITAL STATUS                      0
MONTHLY PREMIUM AUTO                0
MONTHS SINCE LAST CLAIM           633
MONTHS SINCE POLICY INCEPTION       0
NUMBER OF OPEN COMPLAINTS         633
NUMBER OF POLICIES                  0
POLICY TYPE                         0
POLICY                              0
RENEW OFFER TYPE                    0
SALES CHANNEL                       0
TOTAL CLAIM AMOUNT                  0
VEHICLE CLASS                     622
VEHICLE SIZE                      622
VEHICLE TYPE                     5482
dtype: int64

In [24]:
marketing_data[marketing_data['STATE'].isna()==True].head(60)
#checking how NaN values seem to be distributed

UNNAMED: 0 CUSTOMER STATE  CUSTOMER LIFETIME VALUE RESPONSE  COVERAGE  \
5             5  HO81102   NaN              4745.181764      NaN     Basic   
11           11  GE63585   NaN              2420.711888      NaN     Basic   
58           58  BR69566   NaN             14019.833140      NaN     Basic   
62           62  UF44021   NaN              2108.493744      NaN     Basic   
126         126  TE60343   NaN              4716.264294      NaN     Basic   
131         131  QK24293   NaN              6265.343299      NaN  Extended   
155         155  JG35886   NaN              4111.888285      NaN   Premium   
168         168  FV22585   NaN              6960.920371      NaN  Extended   
169         169  AU23837   NaN              5010.379310      NaN     Basic   
190         190  VZ18907   NaN             21991.615980      NaN     Basic   
226         226  VK57280   NaN              4727.354548      NaN     Basic   
251         251  EA52579   NaN              7687.296987      NaN     Basic   
270         270  KX23516   NaN              4455.843406      NaN  Extended   
304         304  PP39078   NaN              5623.611187      NaN     Basic   
305         305  WZ83320   NaN              9061.057741      NaN   Premium   
321         321  WQ85760   NaN              4260.335730      NaN     Basic   
323         323  GF70857   NaN              2933.960426      NaN  Extended   
329         329  OE15600   NaN              2544.969754      NaN     Basic   
334         334  AV76907   NaN              4282.947971      NaN     Basic   
347         347  WT71621   NaN              3280.457296      NaN  Extended   
385         385  OQ61223   NaN             17497.522010      NaN     Basic   
403         403  ET47892   NaN             12280.766610      NaN   Premium   
404         404  XD19546   NaN              8160.689782      NaN     Basic   
430         430  AB96670   NaN              2393.915369      NaN     Basic   
455         455  PO56357   NaN              5363.528173      NaN     Basic   
475         475  MB20764   NaN             17622.133200      NaN   Premium   
478         478  DO98345   NaN             13156.748950      NaN  Extended   
491         491  IE97795   NaN              6878.008403      NaN  Extended   
574         574  WR85740   NaN              8149.769437      NaN     Basic   
595         595  TN17395   NaN             22103.507200      NaN     Basic   
597         597  EP64069   NaN              5138.244998      NaN     Basic   
605         605  OD18225   NaN              6631.665538      NaN  Extended   
648         648  RS29507   NaN             10395.709390      NaN  Extended   
662         662  YS94121   NaN              5645.396665      NaN     Basic   
670         670  RL51023   NaN              4185.757242      NaN     Basic   
686         686  NQ71171   NaN              5107.071054      NaN     Basic   
690         690  OO11998   NaN              4583.748377      NaN     Basic   
720         720  ON74557   NaN             14651.615300      NaN  Extended   
725         725  JC33752   NaN              6277.844624      NaN  Extended   
726         726  SY86059   NaN              3731.504584      NaN  Extended   
729         729  BV86143   NaN             21034.018150      NaN     Basic   
732         732  QR45101   NaN              6624.611830      NaN     Basic   
740         740  RB63074   NaN              4858.067125      NaN     Basic   
744         744  KH24214   NaN             17857.972330      NaN     Basic   
745         745  PB39857   NaN              5129.608404      NaN  Extended   
764         764  ZY57929   NaN              8169.746083      NaN     Basic   
781         781  GI55977   NaN              2337.351820      NaN     Basic   
785         785  KN13494   NaN              5181.620895      NaN     Basic   
794         794  RZ97514   NaN              7498.824515      NaN     Basic   
796         796  KE61355   NaN              5272.703428      NaN     Basic   
814         814  UU70820   

We cant drop the rows with NaN values because they also have importante information.
They are not a lot of NaN values, so we can check if we can fill some of them


In [25]:
marketing_data['STATE'].value_counts()


California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: STATE, dtype: int64

In [26]:
marketing_data['RESPONSE'].value_counts()

No     8813
Yes    1466
Name: RESPONSE, dtype: int64

In [29]:
marketing_data['NUMBER OF OPEN COMPLAINTS'].value_counts()

0.0    8160
1.0    1145
2.0     414
3.0     324
4.0     166
5.0      68
Name: NUMBER OF OPEN COMPLAINTS, dtype: int64

In [30]:
marketing_data['VEHICLE CLASS'].value_counts()

Four-Door Car    5212
Two-Door Car     2118
SUV              2012
Sports Car        550
Luxury SUV        208
Luxury Car        188
Name: VEHICLE CLASS, dtype: int64

In [31]:
marketing_data['VEHICLE SIZE'].value_counts()

Medsize    7251
Small      1966
Large      1071
Name: VEHICLE SIZE, dtype: int64

In [32]:
marketing_data['VEHICLE TYPE'].value_counts()

A    5428
Name: VEHICLE TYPE, dtype: int64

Wewe could fill some of the NaN values with the average or the most frequent element of a column, but I think it is a better option to fill it more randomly using the data from the previous row, as otherwise it would increase too much some of the values

In [33]:
marketing_data_filled = marketing_data.fillna(method='backfill')

In [34]:
marketing_data_filled.isna().sum()

UNNAMED: 0                       0
CUSTOMER                         0
STATE                            1
CUSTOMER LIFETIME VALUE          0
RESPONSE                         1
COVERAGE                         0
EDUCATION                        0
EFFECTIVE TO DATE                0
EMPLOYMENTSTATUS                 0
GENDER                           0
INCOME                           0
LOCATION CODE                    0
MARITAL STATUS                   0
MONTHLY PREMIUM AUTO             0
MONTHS SINCE LAST CLAIM          0
MONTHS SINCE POLICY INCEPTION    0
NUMBER OF OPEN COMPLAINTS        0
NUMBER OF POLICIES               0
POLICY TYPE                      0
POLICY                           0
RENEW OFFER TYPE                 0
SALES CHANNEL                    0
TOTAL CLAIM AMOUNT               0
VEHICLE CLASS                    0
VEHICLE SIZE                     0
VEHICLE TYPE                     1
dtype: int64

We can check, taking 'STATE' as example, that this method doesnt change our values as mach as filling the most frequent (California would have increased in 631, wich is an increase of 17,7%)

In [35]:
marketing_data['STATE'].value_counts()

California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: STATE, dtype: int64

In [36]:
marketing_data_filled['STATE'].value_counts()

California    3765
Oregon        3085
Arizona       2067
Nevada        1051
Washington     941
Name: STATE, dtype: int64